rfd/xxxx-automatic-azure-server-discovery.md

```diff
@@ -0,0 +1,217 @@
+---
+authors: Andrew Burke (andrew.burke@goteleport.com)
+state: draft
+---
+
+# RFD X - Automatic discovery of Azure servers
+
+## Required Approvers
+Engineering: @jakule && @r0mant
+Product: @klizhentas && @xinding33
+Security: @reedloden
+
+## What
+
+Teleport discovery services will be able to automatically discover and enroll Azure virtual machine
+instances. See [RFD 57](https://github.com/gravitational/teleport/blob/master/rfd/0057-automatic-aws-server-discovery.md)
+(Automatic discovery and enrollment of AWS servers) for the same feature implemented
+for AWS.
+
+## Why
+
+RFD 57 replaced the process of manually installing Teleport on EC2 servers, which
+could be slow for large numbers of instances. This RFD does the same for Azure
+instances.
+
+## Details
+
+### Discovery
+
+Azure discovery will be handled by the new [discovery service](https://github.com/gravitational/teleport/blob/master/rfd/0057-automatic-aws-server-discovery.md#discovery) described in RFD 57.
+
+```yaml
+discovery_service:
+  enabled: "yes"
+  azure:
+    - types: ["vm"]
+      subscriptions: ["<subscription id>"]
+      resource_groups: ["<resource group"]
+      regions: ["westcentralus"]
+      tags:
+        "teleport": "yes"
+      install:
+        join_params:
+          token_name:  "azure-discovery-token" # default value
+        nodename: ${SUBSCRIPTION_ID}_${VM_ID}  # default value
+        script_name: "installer"  # default value
+```
+
+The Teleport discovery service will need a [service principal](https://learn.microsoft.com/en-us/cli/azure/create-an-azure-service-principal-azure-cli?view=azure-cli-latest) with a role that includes the `Microsoft.Compute/virtualMachines/read`
+permission to list virtual machines via the Go Azure SDK.
+
+New Azure nodes will be discovered periodically on a 5 minute timer, as new
+nodes are found they will be added to the teleport cluster.
+
+In order to avoid attempting to reinstall Teleport on top of an instance where it is
+already present, the generated Teleport config will match against the node name using
+the Azure subscription ID and the virtual machine ID by default. This can be overridden
+by specifying a node name in the join params.
+
+```json
+{
+  "kind": "node",
+  "version": "v2",
+  "metadata": {
+    "name": "${SUBSCRIPTION}-${VM_ID}",
+    "labels": {
+      "env": "example",
+      "teleport.dev/discovered-node": "yes",
+      "teleport.dev/discovered-by": "${DISCOVER_NODE_UUID}",
+      "teleport.dev/origin": "cloud",
+      "teleport.dev/region": "westcentralus",
+      "teleport.dev/subscriptionId": "88888888-8888-8888-8888-888888888888"
+    }
+  },
+  "spec": {
+    "public_addr": "...",
+    "hostname": "azurexyz"
+  }
+}
+```
+
+### Agent installation
+
+In order to install the Teleport agent on Azure virtual machines, Teleport will serve an
+install script at `/webapi/scripts/{installer-resource-name}`. Installer scripts will
```

Comment:
Is this going to use the same default installer script as AWS discovery?

Reply:
I made a separate one while testing, but there's no reason we couldn't have one default installer that handles both.